pipeline/trusted-resources/pkg/trustedtask/trustedtaskrun.go

```diff
@@ -85,141 +81,79 @@ func (tr *TrustedTaskRun) Validate(ctx context.Context) (errs *apis.FieldError)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "taskRef")
 	}
-
-	if tr.Spec.TaskRef != nil {
-		if tr.Spec.TaskRef.Bundle != "" {
-			logger.Info("Verifying OCI Bundle")
-			if err := VerifyTaskOCIBundle(ctx, tr.Spec.TaskRef.Bundle, verifier, signature, k8sclient); err != nil {
-				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
-			}
-			return nil
-		}
-
-		ts, err := tektonClient.TektonV1beta1().Tasks(tr.Namespace).Get(ctx, tr.Spec.TaskRef.Name, metav1.GetOptions{})
-		if err != nil {
-			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
-		}
-		if ts.Name != "" {
-			logger.Info("Verifying TaskRef")
-			if err := VerifyInterface(ctx, &ts.Spec, verifier, signature); err != nil {
-				if err != nil {
-					return apis.ErrGeneric(err.Error(), "spec", "taskRef")
-				}
-			}
-			return nil
-		}
+	resolvedTask, err := fn(ctx, tr.Spec.TaskRef.Name)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "taskRef")
 	}
 
-	return nil
-}
-
-func verifier(
-	ctx context.Context,
-	annotations map[string]string,
-	k8sclient kubernetes.Interface,
-) (signature.Verifier, error) {
-	if annotations[kmsAnnotation] != "" {
-		// Fetch key from kms.
-		return kms.Get(ctx, annotations[kmsAnnotation], crypto.SHA256)
-	} else {
-		cosignPublicKeypath := secretPath
-		// Overwrite the path if set in configmap.
-		cm, err := k8sclient.CoreV1().ConfigMaps(system.Namespace()).Get(ctx, signingConfigMap, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
-		}
-		if cm.Data["signing-secret-path"] != "" {
-			cosignPublicKeypath = cm.Data["signing-secret-path"]
-		}
-		return cosignsignature.LoadPublicKey(ctx, cosignPublicKeypath)
+	task, signature, err := prepareTask(resolvedTask)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "taskRef")
 	}
-}
 
-func VerifyInterface(
-	ctx context.Context,
-	taskspec interface{},
-	verifier signature.Verifier,
-	signature []byte,
-) (errs *apis.FieldError) {
-	ts, err := json.Marshal(taskspec)
+	verifier, err := verifier(ctx, task.ObjectMeta.Annotations, k8sclient)
 	if err != nil {
-		return apis.ErrGeneric(err.Error(), "taskSpec")
+		return apis.ErrGeneric(err.Error(), "taskRef")
 	}
 
-	h := sha256.New()
-	h.Write(ts)
-
-	if err := verifier.VerifySignature(bytes.NewReader(signature), bytes.NewReader(h.Sum(nil))); err != nil {
-		return apis.ErrGeneric(err.Error(), "taskSpec")
+	if err := VerifyInterface(ctx, task, verifier, signature); err != nil {
+		return apis.ErrGeneric(err.Error(), "taskRef")
 	}
 
 	return nil
 }
 
-func VerifyTaskOCIBundle(
-	ctx context.Context,
-	bundle string,
-	verifier signature.Verifier,
-	signature []byte,
-	k8sclient kubernetes.Interface,
-) (errs *apis.FieldError) {
+// prepareTask will convert the taskobject to task and extract the signature.
+func prepareTask(t v1beta1.TaskObject) (v1beta1.Task, []byte, error) {
+	task := v1beta1.Task{}
+	task.TypeMeta = metav1.TypeMeta{
+		APIVersion: "tekton.dev/v1beta1",
```

Comment:
Nothing to do for this PR - but something to think about for the future is how we might support v1alpha types as well :thinking:

Reply:
sure, I think that is doable